Run the following Python code to split the dataset into training and testing sets. What is the accuracy of the classifier on the test set? (Use random seed = 42)

In [2]:
import numpy as np
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier

# Load dataset
iris = load_iris()
X, y = iris.data, iris.target

# Split dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Train a Decision Tree
model = DecisionTreeClassifier(random_state=42)
model.fit(X_train, y_train)

# Evaluate model
accuracy = model.score(X_test, y_test)
print("Accuracy:", accuracy)


Accuracy: 0.9777777777777777


Write a Python code to perform K-Means clustering on the following dataset. Use a random seed of 50 and number of clusters = 3. What are the cluster centers?

In [11]:
import numpy as np
from sklearn.cluster import KMeans

# Generate dataset
np.random.seed(100)
X = np.random.rand(100, 2)

# Perform K-Means
kmeans = KMeans(n_clusters=3, random_state=50)
kmeans.fit(X)

# Get cluster centers
centers = kmeans.cluster_centers_
print("Cluster Centers:", centers)


Cluster Centers: [[0.1968126  0.48053706]
 [0.70233438 0.79859285]
 [0.66173405 0.21879851]]


Using the dataset provided below, compute the Silhouette score after performing K-Means clustering. Use 3 clusters and a random seed of 2 for reproducibility. What is the Silhouette score, and how would you interpret it based on the general rule of thumb?

In [21]:
import numpy as np
from sklearn.datasets import make_blobs
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score

# Generate synthetic data
X, _ = make_blobs(n_samples=300, centers=3, cluster_std=1.0, random_state=2)

# Perform K-Means clustering
kmeans = KMeans(n_clusters=3, random_state=42)
labels = kmeans.fit_predict(X)

# Calculate Silhouette score
sil_score = silhouette_score(X, labels)
print("Silhouette Score:", sil_score)


Silhouette Score: 0.5732710015800855



### Interpretation
Interpret the score using the rule of thumb:  
- Silhouette score above 0.5 → good clustering  
- Silhouette score between 0.25 and 0.5 → fair clustering  
- Silhouette score below 0.25 → poor clustering  

A Silhouette score of 0.573 indicates a good clustering. Since it is above 0.5, the clusters are reasonably well-separated and compact, though not as optimal as in the case of a higher score (e.g., above 0.7). The data points within each cluster are relatively similar to each other, and there is some distinction between different clusters, but there may be slight overlap or less cohesion between certain clusters.

This clustering can still be considered effective but may benefit from further refinement.